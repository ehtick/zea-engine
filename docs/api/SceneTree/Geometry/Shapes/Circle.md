<a name="Circle"></a>

### Circle 
A class for generating a circle shape.


**Extends**: <code>Lines</code>  
<a name="new_Circle_new"></a>

### new Circle
Create a circle.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| radius | <code>number</code> | <code>1</code> | The radius of the circle. |
| numSegments | <code>number</code> | <code>32</code> | The number of segments. |

