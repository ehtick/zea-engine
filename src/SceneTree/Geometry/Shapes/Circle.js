import { Lines } from '../Lines.js'

import { NumberParameter } from '../../Parameters'
import { sgFactory } from '../../SGFactory.js'

/** A class for generating a circle shape.
 * @extends Lines
 */
class Circle extends Lines {
  /**
   * Create a circle.
   * @param {number} radius - The radius of the circle.
   * @param {number} numSegments - The number of segments.
   */
  constructor(radius = 1.0, numSegments = 32) {
    super()

    if (isNaN(radius) || isNaN(numSegments))
      throw new Error('Invalid geom args')

    this.__radius = this.addParameter(new NumberParameter('Radius', radius))
    this.__radius.valueChanged.connect(this.__resize.bind(this))
    this.__numSegments = this.addParameter(
      new NumberParameter(
        'NumSegments',
        numSegments >= 3 ? numSegments : 3,
        [3, 200],
        1
      )
    )
    this.__numSegments.valueChanged.connect(this.__rebuild.bind(this))
    this.__rebuild()
  }

  /**
   * The __rebuild method.
   * @private
   */
  __rebuild() {
    const segs = this.__numSegments.getValue()
    this.setNumVertices(segs)
    this.setNumSegments(segs)
    for (let i = 0; i < segs; i++) this.setSegment(i, i, (i + 1) % segs)
    this.__resize(-1)
    this.geomDataTopologyChanged.emit()
  }

  /**
   * The __resize method.
   * @param {number} mode - The mode value.
   * @private
   */
  __resize(mode) {
    const radius = this.__radius.getValue()
    const segs = this.__numSegments.getValue()
    const angle = (Math.PI * 2.0) / segs
    for (let i = 0; i < segs; i++)
      this.getVertex(i).set(
        Math.sin(angle * i) * radius,
        Math.cos(angle * i) * radius,
        0.0
      )
    this.setBoundingBoxDirty()
    if (mode != -1) this.geomDataChanged.emit()
  }
}
sgFactory.registerClass('Circle', Circle)
export { Circle }
