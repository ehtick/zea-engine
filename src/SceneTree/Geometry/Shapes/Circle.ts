import { NumberParameter } from '../../Parameters/index'
import { Registry } from '../../../Registry'
import { ProceduralLines } from './ProceduralLines'
import { Vec3Attribute } from '../Vec3Attribute'

/**
 * A class for generating a circle shape using line segments.
 *
 * ```
 * const circle = new Circle(2.2, 12)
 * ```
 *
 * **Parameters**
 * * **Radius(`NumberParameter`):** Radius of the circle.
 * * **Angle(`NumberParameter`):** Number of segments used to build the circle.
 * * **NumSegments(`NumberParameter`):** Segments angle in radiants.
 *
 * @extends {ProceduralLines}
 */
class Circle extends ProceduralLines {
  protected __angle: NumberParameter
  protected __numSegments: NumberParameter
  protected __radius: NumberParameter
  topologyParams: string[]

  /**
   * Creates an instance of Circle.
   * @param {number} radius - The radius of the circle.
   * @param {number} numSegments - The number of segments.
   * @param {number} angle - Arc segments angle(radians)
   */
  constructor(radius = 1.0, numSegments = 32, angle = Math.PI * 2) {
    super()
    this.topologyParams = []
    if (isNaN(radius) || isNaN(numSegments)) throw new Error('Invalid geom args')

    this.__radius = this.addParameter(new NumberParameter('Radius', radius)) as NumberParameter
    this.__angle = this.addParameter(new NumberParameter('Angle', angle)) as NumberParameter
    this.__numSegments = this.addParameter(
      new NumberParameter('NumSegments', numSegments >= 3 ? numSegments : 3, [3, 200], 1)
    ) as NumberParameter

    this.topologyParams.push('NumSegments')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const segs = this.__numSegments.getValue() || 32
    this.setNumVertices(segs)
    const angle = this.__angle.getValue() || Math.PI * 2
    const arc = angle < Math.PI * 2
    if (arc) this.setNumSegments(segs - 1)
    else this.setNumSegments(segs)
    for (let i = 0; i < (arc ? segs - 1 : segs); i++) this.setSegmentVertexIndices(i, i, (i + 1) % segs)
    this.resize()
  }

  /**
   * The resize method.
   * @private
   */
  resize(): void {
    const radius = this.__radius.getValue() || 1.0
    const segs = this.__numSegments.getValue() || 32
    const angle = this.__angle.getValue() || Math.PI * 2
    const step = angle / segs
    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    if (positions) {
      for (let i = 0; i < segs; i++) {
        positions.getValueRef(i).set(Math.cos(step * i) * radius, Math.sin(step * i) * radius, 0.0)
      }
    }
  }
}

Registry.register('Circle', Circle)
export { Circle }
