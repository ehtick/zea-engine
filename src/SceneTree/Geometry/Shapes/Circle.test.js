import { Circle } from './Circle'

describe('Circle', () => {
  it('tests default parameters', () => {
    const circle = new Circle()

    expect(circle.getParameter('Radius').getValue()).toBe(1.0)
    expect(circle.getParameter('Angle').getValue()).toBe(Math.PI * 2)
    expect(circle.getParameter('Sides').getValue()).toBe(32)
  })

  it('updates parameters', () => {
    const circle = new Circle()
    circle.getParameter('Radius').setValue(2)
    circle.getParameter('Angle').setValue(Math.PI)
    circle.getParameter('Sides').setValue(64)

    expect(circle.getParameter('Radius').getValue()).toBe(2)
    expect(circle.getParameter('Angle').getValue()).toBe(Math.PI)
    expect(circle.getParameter('Sides').getValue()).toBe(64)
  })

  it('saves to JSON (serialization).', () => {
    const circle = new Circle(2, 6, Math.PI * 2)
    const outputJSON = circle.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const circle = new Circle()
    const expectedOutput = {
      params: {
        Angle: { value: 6.283185307179586 },
        Sides: {
          range: [3, 200],
          step: 1,
          value: 6,
        },
        Radius: { value: 2 },
      },
      type: 'Circle',
      vertexAttributes: {},
    }
    circle.fromJSON(expectedOutput)

    const newCircle = new Circle(2, 6, Math.PI * 2)
    expect(circle.toJSON()).toEqual(newCircle.toJSON())
  })
})
