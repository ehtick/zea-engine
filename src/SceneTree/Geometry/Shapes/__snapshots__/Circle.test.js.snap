// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle saves to JSON (serialization). 1`] = `
Object {
  "params": Object {
    "Angle": Object {
      "value": 6.283185307179586,
    },
    "Radius": Object {
      "value": 2,
    },
    "Sides": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 6,
    },
  },
  "type": "Circle",
  "vertexAttributes": Object {},
}
`;
